pinecone/index_connection.go

```diff
@@ -0,0 +1,358 @@
+package pinecone
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"github.com/pinecone-io/go-pinecone/internal/gen/data"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+	"log"
+)
+
+type IndexConnection struct {
+	apiKey     string
+	dataClient *data.VectorServiceClient
+	grpcConn   *grpc.ClientConn
+}
+
+func newIndexConnection(apiKey string, host string) (*IndexConnection, error) {
+	config := &tls.Config{}
+	target := fmt.Sprintf("%s:443", host)
+	conn, err := grpc.Dial(
+		target,
+		grpc.WithTransportCredentials(credentials.NewTLS(config)),
+		grpc.WithAuthority(target),
+		grpc.WithBlock(),
+	)
+
+	if err != nil {
+		log.Fatalf("fail to dial: %v", err)
+		return nil, err
+	}
+
+	dataClient := data.NewVectorServiceClient(conn)
+
+	idx := IndexConnection{apiKey: apiKey, dataClient: &dataClient, grpcConn: conn}
+	return &idx, nil
+}
+
+func (idx *IndexConnection) Close() error {
+	err := idx.grpcConn.Close()
+	return err
+}
+
+type UpsertVectorsRequest struct {
```

Comment:
**question:** Some structs are defined in-line in this file and some are in `models.go`, would it make sense to keep everything in one place or is it more reasonable to define these things near where they're used?

Reply:
It's sudo-arbitrary. The exported structs defined here are the Request and Response structs for exported actions. The data structs are in `models.go`. I guess because those have a higher probability of being reused outside of index connection?

This is something that can be re-organized in the future without breaking uses of the SDK. Everything is exported to the `pinecone` package regardless of filename.